### PR TITLE
#1364 Bluetooth module

### LINKED ISSUE #1364: Bluetooth low energy module
## Is your feature request related to a problem? Please describe.

Need to add the BLE module created for vaccine module

## Describe the solution you'd like

Add:

https://github.com/openmsupply/mobile/tree/vaccine-development/android/app/src/main/java/com/msupplymobile/Ble

## Describe alternatives you've considered

N/A

## Additional context

N/A

### PR DESCRIPTION
**Pull request by joshxg**
Fixes #1364 

## Change summary

- Just a port of the blue tooth module from the previous develop branch.
- I can see a few places where I'd like to improve a few things, but going to leave it for now - it's working and sending packets. I
- I think improvements can be done when we make a npm package?

## Testing

N/A [Will do tests when the feature branch is complete]

### Related areas to think about

Little pic maybe helpful, maybe not. As above there are improvements needed/could do but the general idea of having specific 'device' concrete classes which relate to a specific physical sensor which is composed of a parser for a specific physical sensor implementation

https://drive.google.com/file/d/1nhtpZNPP1AWvaGzPhOVHqGETEOvAO0Mp/view?usp=sharing

### CODE REVIEW

File: android/app/src/main/java/com/msupplymobile/Ble/BleDevice.java [hunk 1]
```diff
@@ -0,0 +1,77 @@
...
+    
+    protected BleDeviceListener bleDeviceListener;
+    protected ScanResult scanResult;
+
+    public BleDevice(ScanResult scanResult){
+        this.scanResult = scanResult;
+    }
+
+    /**  Registers a listener. Currently supports just one listener */
+    public void registerListener(BleDeviceListener bleLDeviceistener){
```
Comment (andreievg) [inline android/app/src/main/java/com/msupplymobile/Ble/BleDevice.java]: a bit pedantic with spelling

Comment (mark-prins) [reply android/app/src/main/java/com/msupplymobile/Ble/BleDevice.java]: wow, good spotting!

File: android/app/src/main/java/com/msupplymobile/Ble/BlueMaestroDevice.java [hunk 2]
```diff
@@ -0,0 +1,200 @@
...
+     */
+    @Override
+    public WritableMap toObject(){
+        WritableMap asObject = this.parseAdvertisement();
+        asObject.putString("name", getName());
+        asObject.putString("macAddress", getAddress());
+        asObject.putInt("rssi", getRssi());
+        if (command.equals("*logall")){
+            asObject.putArray("logs", this.parseLogs());
+        }
```
Comment (andreievg) [inline android/app/src/main/java/com/msupplymobile/Ble/BlueMaestroDevice.java:85]: i think if command != "" we should add "commandString" or something analog

File: android/app/src/main/java/com/msupplymobile/Ble/BleManager.java [hunk 3]
```diff
@@ -0,0 +1,202 @@
...
+        return "BleManager";
+    }
+
+    /** 
+     * Callback for handling commands sent to devices,
+     * whether it was succcesful, or not.
+     * Overrides BleDeviceListener
+     */
+    @Override
+    public void BleCallback(boolean result){
```
Comment (andreievg) [inline android/app/src/main/java/com/msupplymobile/Ble/BleManager.java:52]: little bit iffy about this one, the only 'better' way i can think of here is to just return the device and have it as a single object in 'data'. So different return to JS if we are scanning for devices ({devices: [devices]}), or getting result of command for one device ({device: {single device info}} <- actually i think that would make more sense. what cha reckon ?

**Pull request merged**